controllers/capabilities/tenant_controller.go

```diff
@@ -95,119 +107,116 @@ func (r *TenantReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		reqLogger.V(1).Info(string(jsonData))
 	}
 
-	masterAccessToken, err := r.fetchMasterCredentials(tenantR)
-	if err != nil {
-		reqLogger.Error(err, "Error fetching master credentials secret")
-		// Error reading the object - requeue the request.
-		return ctrl.Result{}, err
+	// reconcileErr can be an error of different types - generic k8s err, isInvalid, waitError, isOrphaned
+	// different errors are handled differently when it comes to requeueing the reconciliation loop
+	statusReconciler, reconcileErr := r.specReconciler(tenantR, reqLogger)
+
+	// statusUpdateError can be an error when making the update call to the resource
+	res, statusUpdateErr := statusReconciler.Reconcile()
+	if statusUpdateErr != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update tenantCR: %w", statusUpdateErr)
 	}
 
-	insecureSkipVerify := controllerhelper.GetInsecureSkipVerifyAnnotation(tenantR.GetAnnotations())
-	portaClient, err := controllerhelper.PortaClientFromURLString(tenantR.Spec.SystemMasterUrl, masterAccessToken, insecureSkipVerify)
-	if err != nil {
-		reqLogger.Error(err, "Error creating porta client object")
-		// Error reading the object - requeue the request.
-		return ctrl.Result{}, err
+	// Requeue if either object or status were updated
+	if res.Requeue {
+		return reconcile.Result{}, nil
 	}
 
-	// Tenant has been marked for deletion
-	if tenantR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
-		existingTenant, err := controllerhelper.FetchTenant(tenantR.Status.TenantId, portaClient)
-		if err != nil {
-			return ctrl.Result{}, err
+	if reconcileErr != nil {
+		// On validation error, do not retry since there's something wrong with the spec
+		if helper.IsInvalidSpecError(reconcileErr) {
+			reqLogger.Info("ERROR", "spec validation error", reconcileErr)
+			r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "Invalid Spec", "%v", reconcileErr)
+			return ctrl.Result{}, nil
 		}
 
-		// delete tenantCR if tenant is present in 3scale
-		if existingTenant != nil {
-			// do not attempt to delete tenant that is already scheduled for deletion
-			if existingTenant.Signup.Account.State != scheduledForDeletionState {
-				err := portaClient.DeleteTenant(tenantR.Status.TenantId)
-				if err != nil {
-					r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "Failed to delete tenant", "%v", err)
-					return ctrl.Result{}, err
-				}
-			} else {
-				reqLogger.Info("Removing tenant CR - tenant is already scheduled for deletion", "tenantID", existingTenant.Signup.Account.ID)
-			}
+		// On wait error, retry
+		if helper.IsWaitError(reconcileErr) {
+			reqLogger.Info("ERROR", "wait error", reconcileErr)
+			return ctrl.Result{Requeue: true}, nil
 		}
 
-		controllerutil.RemoveFinalizer(tenantR, tenantFinalizer)
-		err = r.UpdateResource(tenantR)
-		if err != nil {
-			return ctrl.Result{}, err
+		// On orphan error, do not retry
+		if helper.IsOrphanSpecError(reconcileErr) {
```

Comment:
The only time orphaned is used when tenant CR is marked for deletion but the tenant is already marked for deletion in 3scale as well. IMO this is not a validation error but is also an error that should not re-trigger the reconciler. 
We used to delete the CR however I've removed this logic since to me it seemed that the tenant/3scale must have been some weird state for this to happen. Now, I think we might also have the issue when tenant delete is successful but reconciler fails somewhere else (after deleting tenant from 3scale) - in that case we should remove CR.. any suggestions?

Reply:
Not sure I follow. 

If the 3scale tenant is either deleted (not found) or marked for deletion, it is safe to remove the finalizer and let the Tenant CR be removed.

Reply:
Ok, I will remove the CR if tenant is already marked for deletion.

Reply:
Updated, thank you